realm/src/androidTest/java/io/realm/RealmInterprocessTest.java

```diff
@@ -0,0 +1,260 @@
+    private Messenger serviceMessenger;
+    private Messenger receiverMessenger;
+    private CountDownLatch serviceLatch;
+    private final ServiceConnection serviceConnection = new ServiceConnection() {
+        @Override
+        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
+            serviceMessenger = new Messenger(iBinder);
+            serviceLatch.countDown();
+        }
+
+        @Override
+        public void onServiceDisconnected(ComponentName componentName) {
+        }
+    };
+
+    // It is necessary to overload this method.
+    // AndroidTestRunner does call Looper.prepare() and we can have a looper in the case. The problem is all the test
+    // cases are running in a single thread!!! And after Looper.quit() called, it cannot start again. That means we
+    // can only have one case in this class LoL.
+    // By overloading this method, we create a new thread and looper to run the real case. And use latch to wait until
+    // it is finished. Then we can get rid of creating thread in the test method, using array to store exception, many
+    // levels of nested code. Make the test case more nature.
+    @Override
+    public void runBare() throws Throwable {
+        final Throwable[] throwableArray = new Throwable[1];
+        final CountDownLatch latch = new CountDownLatch(1);
+        Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Looper.prepare();
+                try {
+                    RealmInterprocessTest.super.runBare();
+                } catch (Throwable throwable) {
+                    throwableArray[0] = throwable;
+                } finally {
+                    latch.countDown();
+                }
+            }
+        });
+
+        thread.start();
+        try {
+            latch.await();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        if (throwableArray[0] != null) {
+            throw throwableArray[0];
+        }
+    }
+
+    // Helper handler to make it easy to interact with service process.
+    // This handler run in newly created thread's Looper. When writing the test case, remember to call
+    // Looper.loop() to start handling message.
+    // Pass the first thing you want to run to the constructor which will be posted to the beginning
+    // of the message queue.
+    // Write the comments of the test case like this:
+    // A-Z means steps running from service process.
+    // 1-9xx means steps running from the main process.
+    @SuppressLint("HandlerLeak") // SuppressLint bug, doesn't work
+    private class InterprocessHandler extends Handler {
+        // Timeout Watchdog. In case the service crashed or expected response is not returned.
+        // It is very important to feed the dog after the expected message arrived.
+        private final int timeout = 5000;
+        private boolean timeoutFlag = true;
```

Comment:
`timeoutFlag` needs to be `volatile` it's accessed from different thread (_clearTimeoutFlag_ & _InterprocessHandler_)